src/libraries/System.Text.Json/gen/Helpers/RoslynExtensions.cs

```diff
@@ -0,0 +1,255 @@
+            }
+        }
+
+        public static bool IsGenericTypeDefinition(this ITypeSymbol type)
+            => type is INamedTypeSymbol { IsGenericType: true } namedType && SymbolEqualityComparer.Default.Equals(namedType, namedType.ConstructedFrom);
+
+        public static bool IsNumberType(this ITypeSymbol type)
+        {
+            return type.SpecialType is
+                SpecialType.System_SByte or SpecialType.System_Int16 or SpecialType.System_Int32 or SpecialType.System_Int64 or
+                SpecialType.System_Byte or SpecialType.System_UInt16 or SpecialType.System_UInt32 or SpecialType.System_UInt64 or
+                SpecialType.System_Single or SpecialType.System_Double or SpecialType.System_Decimal;
+        }
+
+        public static bool IsNullableValueType(this ITypeSymbol type, [NotNullWhen(true)] out ITypeSymbol? elementType)
+        {
+            if (type.IsValueType && type is INamedTypeSymbol { OriginalDefinition.SpecialType: SpecialType.System_Nullable_T })
+            {
+                elementType = ((INamedTypeSymbol)type).TypeArguments[0];
+                return true;
+            }
+
+            elementType = null;
+            return false;
+        }
+
+        public static ITypeSymbol[] GetAllTypeArgumentsInScope(this INamedTypeSymbol type)
+        {
+            if (!type.IsGenericType)
+            {
+                return Array.Empty<ITypeSymbol>();
+            }
+
+            var args = new List<ITypeSymbol>();
+            TraverseContainingTypes(type);
+            return args.ToArray();
+
+            void TraverseContainingTypes(INamedTypeSymbol current)
+            {
+                if (current.ContainingType is INamedTypeSymbol parent)
+                {
+                    TraverseContainingTypes(parent);
+                }
+
+                args.AddRange(current.TypeArguments);
+            }
+        }
+
+        public static ITypeSymbol GetMemberType(this ISymbol member)
+        {
+            Debug.Assert(member is IFieldSymbol or IPropertySymbol);
+            return member is IFieldSymbol fs ? fs.Type : ((IPropertySymbol)member).Type;
+        }
+
+        public static bool MemberNameNeedsAtSign(this ISymbol symbol)
+            => SyntaxFacts.GetKeywordKind(symbol.Name) != SyntaxKind.None || SyntaxFacts.GetContextualKeywordKind(symbol.Name) != SyntaxKind.None;
+
+        public static INamedTypeSymbol[] GetSortedTypeHierarchy(this ITypeSymbol type)
+        {
+            if (type is not INamedTypeSymbol namedType)
+            {
+                return Array.Empty<INamedTypeSymbol>();
+            }
+
+            if (type.TypeKind != TypeKind.Interface)
+            {
+                var list = new List<INamedTypeSymbol>();
+                for (INamedTypeSymbol? current = namedType; current != null; current = current.BaseType)
+                {
+                    list.Add(current);
+                }
+
+                return list.ToArray();
+            }
```

Comment:
Is this part of the algorithm the same as the reflection implementation

Reply:
It's much simpler to express compared to topological sorting, so extracting an abstraction out of it isn't worth the hassle in my view.